refactor: optimised `options` by moving all `getters` and `setters` to custom descriptors

sandeepsuryaprasad opened the pull request:
- In `appium/options`, there were tons and tons of python modules with hundreds of `properties`.

- If we carefully observe the `getters` and `setters`, all `getters` are calling `get_capabilites` method and all setter methods are calling `set_capabilites` method (with some exceptions for for attributes).

- I could see that there is lot of boiler plate code in the entire `options` package.

- There is lot scope for optimisation and I have done it by eliminating `properties` by custom data `descriptors`.

- Custom Descriptors are most efficient and ideal in this case where it eliminates multiple tons of `getters` and `setters`.

- Now the `options` folder looks more light weight.

- I am still working on docstings and other lynting stuff for the changes that I have made. I will be updating this PR with final version of the code in next 24 hrs.

- Mean while let me know your thoughts on the changes that I have made.

mykola-mokhnach commented:
I always appreciate PRs that remove more code than add. Although I would still prefer to see a version first where typing works as expected and proper docstrings are present for each capability.

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,102 @@
+
+
+class AdbExecTimeoutOption(SupportsCapabilities):
+    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
```
I assume this property is also redundant. We could simply use `self.__class__.__name__` itself without the `Option` suffix

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,102 @@
+
+
+class AdbOptionsDescriptor:
+    def __init__(self, name: str) -> None:
```
we could pass the class instance here and use `cls is ...` validation inside of the getter and setter instead of hardcoding strings

mykola-mokhnach replied:
eventually I'm not a big fan of keeping descendant-specific logic inside of a super class. The way properties are transformed should only be known (isolated) to a specific option class

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,102 @@
+    def __init__(self, name: str) -> None:
+        self.name = name
+
+    def __get__(self, obj: Any, cls: Any) -> Any:
```
looking forward to see proper type declarations on these args. getattr does not look right in this context

sandeepsuryaprasad replied:
@mykola-mokhnach thanks for taking time to review my PR. I will work on your review comments. Thanks!

sandeepsuryaprasad replied:
> 

```
capability = TypeVar("capability", bound=SupportsCapabilities)

class OptionsDescriptor:
    def __init__(self, name: str) -> None:
        self.name = name

    def __get__(self, obj: capability, cls: type[capability]) -> Any:
        return getattr(obj, "get_capability")(self.name)

    def __set__(self, obj: capability, value: Any) -> capability:
        getattr(obj, "set_capability")(self.name, value)
```
@mykola-mokhnach are the above type hints okay?

mykola-mokhnach commented on test/unit/junit.xml:
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite name="pytest" errors="0" failures="0" skipped="0" tests="122" time="2.924" timestamp="2023-06-22T00:04:16.153394" hostname="Sandeeps-MacBook-Pro.local"><testcase classname="unit.webdriver.app_test.TestWebDriverApp" name="test_reset" time="0.019" /><testcase classname="unit.webdriver.app_test.TestWebDriverApp" name="test_install_app" tim ...
```
this file should not be under source control

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,227 @@
+        if self.name == 'ADB_EXEC_TIMEOUT':
+            value = getattr(obj, 'get_capability')(self.name)
+            return None if value is None else timedelta(milliseconds=value)
+        return getattr(obj, 'get_capability')(self.name)
```
getattr is now redundant because linter knows that the object implements the corresponding interface

sandeepsuryaprasad replied:
> getattr is now redundant because linter knows that the object implements the corresponding interface

I did not quite understand this. Correct me if am wrong. In the descriptor class we have to have `__get__` method as it internally calls `get_capability` method as soon as we do an attribute lookup, which in turn pulls out the value of attribute let's say `ADB_EXEC_TIMEOUT` from the internal dictionary.  So I have to ask `__get__` method to call `get_capability` method in which is implemented in `base.py`, which I am doing it through `getattr` method. I am confused as to how `getattr` is redundant in this case.

sandeepsuryaprasad replied:
> getattr is now redundant because linter knows that the object implements the corresponding interface

I have made this change in the code. Now instead of using `getattr` and `setattr` method I am directly calling `get_capability` and `set_capability` methods using `obj` in `__get__` and `__set__` method.

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,227 @@
+class AdbExecTimeoutOption(SupportsCapabilities):
+    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
+    adb_exec_timeout = AdbOptionsDescriptor("ADB_EXEC_TIMEOUT")
+    """
```
this docustring is not very helpful as it is not going to be shown by IDE for the corresponding property

sandeepsuryaprasad replied:
> this docustring is not very helpful as it is not going to be shown by IDE for the corresponding property

The docstring shows up in IDE. I am using VSCode. Attaching screenshot below.

<img width="837" alt="docstrings" src="https://github.com/appium/python-client/assets/26169602/b2b97351-94e5-4120-9a5f-3501e775814a">

mykola-mokhnach replied:
it looks ok, although is still not ideal as there is no proper typehint for the value, so we don't actually know what valid objects could be assigned to it and what the returned type is

sandeepsuryaprasad replied:
> it looks ok, although is still not ideal as there is no proper typehint for the value, so we don't actually know what valid objects could be assigned to it and what the returned type is

We can add notes and have complete information about the type of object that can be assigned and return value Information. I will do this change.

sandeepsuryaprasad replied:
> it looks ok, although is still not ideal as there is no proper typehint for the value, so we don't actually know what valid objects could be assigned to it and what the returned type is

The below docstring should be good enough. Let me know your comments/thoughts.

<img width="850" alt="docstring" src="https://github.com/appium/python-client/assets/26169602/5a94fc70-78a9-4ddc-b461-ceecd00d7f7a">

mykola-mokhnach replied:
this docstring looks better to me

sandeepsuryaprasad replied:
> this docstring looks better to me

Thank You for your feedback. I will apply the same style to all the other descriptors.

sandeepsuryaprasad replied:
> this docstring looks better to me

This looks even better and more informative about each `gettter` and `setter`.

<img width="998" alt="doc_strings" src="https://github.com/appium/python-client/assets/26169602/aedb4dff-0265-4d8a-ae1d-09bcc7ef55c9">

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,227 @@
+            return None if value is None else timedelta(milliseconds=value)
+        return getattr(obj, 'get_capability')(self.name)
+
+    def __set__(self, obj:C, value: C) -> C:
```
`value` is an instance of `C`

sandeepsuryaprasad replied:
> `value` is an instance of `C`

I have fixed it. Also, I have used single quotes for representing strings.

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,227 @@
+
+from appium.options.common.supports_capabilities import SupportsCapabilities
+
+C = TypeVar('C', bound='SupportsCapabilities')
```
I assume this class needs must be a [generic](https://mypy.readthedocs.io/en/stable/generics.html) one

sandeepsuryaprasad replied:
> I assume this class needs must be a [generic](https://mypy.readthedocs.io/en/stable/generics.html) one

But then what should be the type of `obj` and `cls` in `__get__` and `__set__` method. I think it has to be of type `SupportCapabilities`.

mykola-mokhnach replied:
I actually meant the descriptor class this is needed in order to properly describe the type returned by the getter and set in the setter

sandeepsuryaprasad replied:
> I actually meant the descriptor class. This is needed in order to properly describe the type returned by the getter and set in the setter

Got it. Thanks! I will do this change.

sandeepsuryaprasad replied:
@mykola-mokhnach I have incorporated all the review comments that were given. Request you the review and let me know in case if I need to change anything. Thank You.

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,383 @@
+    def __init__(self, name: str) -> None:
+        self.name = name
+
+    def __get__(self, obj: C, cls: type[C]) -> Any:
```
the returned result must be types and should be based on the generic type

mykola-mokhnach replied:
also please use `from typing import Type`. The declaration of `type[C]` is supported only since Python 3.8

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,383 @@
+            return None if value is None else timedelta(milliseconds=value)
+        return obj.get_capability(self.name)
+
+    def __set__(self, obj: C, value: Any) -> None:
```
the value argument must be typed and should be based on the generic type

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,383 @@
+        self.name = name
+
+    def __get__(self, obj: C, cls: type[C]) -> Any:
+        if self.name == 'ADB_EXEC_TIMEOUT':
```
this logic must not be present in the descriptor classes, but rather incapsulated in corresponding Option classes

sandeepsuryaprasad replied:
> st not be present in

I tried this. I cannot have the logic for descriptor lookup in specific Option class. I have to have both `__get__` and `__set__` method in  descriptor class itself.

```
class Descriptor:
    def __get__(self, obj, cls):
        print("__get__ Descriptor")

    def __set__(self, obj, value):
        print("__set__ Descriptor")


class Option1(Descriptor):
    name = Descriptor()

    def __get__(self, obj, cls):
        print("__get__ Option1")
```
`option = Option1()`
when I do an attribute look-up `option.name` it always calls parent class `__get__` I can't override it.

mykola-mokhnach replied:
yes, the solution above won't work. What I would do is to pass two transformer functions as constructor arguments - one for the setter and one for the getter. By default these could be None, which means the value is get/set as is, otherwise it must be preprocessed by the given transformer(s) first.

sandeepsuryaprasad replied:
```
class AdbOptionsDescriptor(Generic[T]):
    def __init__(self, name: str, fget=None, fset=None) -> None:
        self.name = name
	self.fget = fget
        self.fset = fset

    def __get__(self, obj: C, cls: type[C]) -> Any:
        if self.fget:
            return self.fget(obj)
        return obj.get_capability(self.name)

    def __set__(self, obj: C, value: Any) -> None:
        if self.fset:
            self.fset(obj, value)
        obj.set_capability(self.name, value)


class AdbExecTimeoutOption(SupportsCapabilities):
    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
    
    def getter(self):
        value = self.get_capability("ADB_EXEC_TIMEOUT")
        return None if value is None else timedelta(milliseconds=value)

    def setter(self, value):
        self.set_capability(self.name, int(value.total_seconds() * 1000) if isinstance(value, timedelta) else value)    
    
    adb_exec_timeout = AdbOptionsDescriptor('ADB_EXEC_TIMEOUT', getter, setter)



class AdbPortOption(SupportsCapabilities):
    ADB_PORT = 'adbPort'
    adb_port = AdbOptionsDescriptor('ADB_PORT')
```

In the above code we have two option classes, in the first option class `AdbExecTimeoutOption` we have defined transformation functions `getter` and `setter` and passing it to descriptor class `AdbOptionsDescriptor` as constructor argument. Now the `__get__` and `__set__` methods in the descriptor calls the transformation functions defined in `AdbExecTimeoutOption`.

In the second option class, we have not defined any transformation function. So the `__get__` and `__set__` function calls `get_capability` and `set_capability` functions directly.

Is this how you want me to extend this idea to other option class?

mykola-mokhnach replied:
Close, but not exactly that:

```python

def noop(x: T) -> T:
    return x

class OptionsDescriptor(Generic[R, V]):
    def __init__(self, name: str, tget: Optional[Callable[[Any], R]] = None, tset: Optiona[Callable[[V], Any]]) -> None:
        self.name = name
	self.tget = tget
        self.tset = tset

    def __get__(self, obj: C, cls: Type[C]) -> R:
        transformer = self.tget or noop
        return transformer(obj.get_capability(self.name))

    def __set__(self, obj: C, value: V) -> None:
        transformer = self.tset or noop
        obj.set_capability(self.name, transformer(valule))


class AdbExecTimeoutOption(SupportsCapabilities):    
    def transform_get(self, value: Union[timedelta, int]) -> Optional[timedelta]:
        return None if value is None else timedelta(milliseconds=value)

    def transform_set(self, value: Union[timedelta, int]) -> int:
        return int(value.total_seconds() * 1000) if isinstance(value, timedelta) else value
    
    adb_exec_timeout = OptionsDescriptor[Optional[timedelta], Union[Timedelta, Int]](
        'adbExecTimeout', 
        self.transform_get, 
        self.transform_set,
    )


class AdbPortOption(SupportsCapabilities):
    adb_port = OptionsDescriptor('adbPort')

```

sandeepsuryaprasad replied:
```
class AdbOptionsDescriptor(Generic[T]):
    def __init__(self, name: str, fget=None, fset=None) -> None:
        self.name = name
        self.fget = fget
        self.fset = fset

    def __get__(self, obj: C, cls: type[C]) -> Any:
        value = obj.get_capability(self.name)
        if self.fget:
            # transform the value and then return
            return self.fget(obj, value)
        return value

    def __set__(self, obj: C, value: Any) -> None:
        if self.fset:
            # transform the value before setting
            value = self.fset(obj, value)
        obj.set_capability(self.name, value)


class AdbExecTimeoutOption(SupportsCapabilities):
    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
    
    def getter(self, value):
        return None if value is None else timedelta(milliseconds=value)

    def setter(self, value):
        return int(value.total_seconds() * 1000) if isinstance(value, timedelta) else value
    
    adb_exec_timeout = AdbOptionsDescriptor('ADB_EXEC_TIMEOUT', getter, setter)

class AdbPortOption(SupportsCapabilities):
    ADB_PORT = 'adbPort'
    adb_port = AdbOptionsDescriptor('ADB_PORT')
```

sandeepsuryaprasad replied:
If `fget` and `fset` is passed in, then transform the `value` and call `get_capability` and `set_capability`. Otherwise directly call `get_capability` and `set_capability` directly by passing the original `value`.  Just avoided `noop` function.  If you are fine with this, I will apply this change to all options. Thank You.

mykola-mokhnach replied:
these transformers could also be repeated for some caps (e.g. there are several values that accept and return tiemdeltas ), so you may decide extracting them into a shared module to avoid unnecessary repetition

sandeepsuryaprasad replied:
Exactly. I have seen lot of transformers which returns `timedeltas` which has repetitive code. Yes we can move those transformers which are repetitive into separate module. But before that I am gonna make the above changes first and then work on separate module for repetitive transformers.

sandeepsuryaprasad replied:
I have modified the descriptors to accept transformation functions. Moved the descriptor to separate module under `options/transformers.py` where in we have descriptor class `OptionsDescriptor`, `transform_get` and `transform_set` functions which is imported to the specific option class.  If the transformation is to specific class like in `other_options.py/CommandTimeoutsOption`, the transformation function is written in specific class  (local, cannot be re-used).

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,39 @@
+            value = self.tset(obj, value)
+        obj.set_capability(self.name, value)
+
+def transform_get(self, value):
```
why do we need the `self` argument?

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,39 @@
+    """Transfroms the value into timedelta"""
+    return None if value is None else timedelta(milliseconds=value)
+
+def transform_set(self, value):
```
These transformer function names because they are not universal functions, but only intended to transform durations. Other capabilities might need other transformers.

sandeepsuryaprasad replied:
Fixed

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,39 @@
+    If transformation method is passed, get_capability and set_capability:
+    method will be called with transformed value
+    """
+    def __init__(self, name: str, tget=None, tset=None) -> None:
```
Please add argument type definitions

sandeepsuryaprasad replied:
Fixed.

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,39 @@
+            return self.tget(obj, value)
+        return value
+
+    def __set__(self, obj: C, value: Any) -> None:
```
value must be properly typed

sandeepsuryaprasad replied:
Fixed

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,39 @@
+        self.tget = tget
+        self.tset = tset
+
+    def __get__(self, obj: C, cls: Type[C]) -> Any:
```
the returned result must be properly typed

sandeepsuryaprasad replied:
Fixed

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,39 @@
+    def __set__(self, obj: C, value: Any) -> None:
+        if self.tset:
+            # transform the value before setting
+            value = self.tset(obj, value)
```
this is not a good pattern to reassign function argument values

sandeepsuryaprasad replied:
fixed it by giving different name to returned value

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,39 @@
+            # transform the value before setting
+            value = self.tset(obj, value)
+        obj.set_capability(self.name, value)
+
```
transformer functions must be properly typed

sandeepsuryaprasad replied:
Fixed

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,39 @@
+T = TypeVar('T')
+C = TypeVar('C', bound='SupportsCapabilities')
+
+class OptionsDescriptor(Generic[T]):
```
OptionsDescriptor class is some base entity. It must not be in the same module where transformers are defined

sandeepsuryaprasad replied:
Fixed.

mykola-mokhnach commented on appium/options/ios/xcuitest/other/other_options.py:
```diff
@@ -0,0 +1,155 @@
+
+class UseJsonSourceOption(SupportsCapabilities):
+    USE_JSON_SOURCE = 'useJSONSource'
+    use_json_source = OptionsDescriptor('USE_JSON_SOURCE')
```
are you sure the correct value is passed there? We need the option name in capabilities, not the class property name.

sandeepsuryaprasad replied:
> are you sure the correct value is passed there? We need the option name in capabilities, not the class property name.

Yes, it is correct. I just verified with original `getter` and `setter` method.

sandeepsuryaprasad replied:
@mykola-mokhnach I have incorporated all the review comments. Could you please re-check once let me know your comments.

mykola-mokhnach commented on appium/options/base_options.py:
```diff
@@ -0,0 +1,34 @@
+
+    def __get__(self, obj: C, cls: Type[C]) -> RT:
+        value = obj.get_capability(self.name)
+        if self.tget:
```
prefer to use `is` comparison with None: https://stackoverflow.com/questions/3257919/what-is-the-difference-between-is-none-and-none

sandeepsuryaprasad replied:
I have changed the conditional to use `is`. now in conditional I am using `if not self.tget is None`

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,367 @@
+class AdbExecTimeoutOption(SupportsCapabilities):
+    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
+    
+    _transform_duration_get = DurationTransformer.transform_duration_get
```
is it necessary to reassign these functions?

sandeepsuryaprasad replied:
I could do that or I can directly access.
`unlock_success_timeout = OptionsDescriptor('UNLOCK_SUCCESS_TIMEOUT', DurationTransformer.transform_duration_get, DurationTransformer.transform_duration_set)`

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,367 @@
+    _transform_duration_get = DurationTransformer.transform_duration_get
+    _transform_duration_set = DurationTransformer.transform_duration_set
+    
+    adb_exec_timeout = OptionsDescriptor('ADB_EXEC_TIMEOUT', _transform_duration_get, _transform_duration_set)
```
I am still not convinced about passing uppercase characters as `name` argument. It is used as an argument for set/get_capability and there we expect a camel-case cap name

mykola-mokhnach replied:
generic types must properly be provided for each descriptor. You can run `make mypy` to verify typing locally

sandeepsuryaprasad replied:
> I am still not convinced about passing uppercase characters as `name` argument. It is used as an argument for set/get_capability and there we expect a camel-case cap name

Yes.. Thanks for identifying this bug. I went through the `set_capability` once again and understood that we are creating a key with camelcase in `_caps` dictionary

```
    def set_capability(self: T, name: str, value: Any) -> T:
        w3c_name = name if name in self.W3C_CAPABILITY_NAMES or ':' in name else f'{APPIUM_PREFIX}{name}'
        if value is None:
            if w3c_name in self._caps:
                del self._caps[w3c_name]
        else:
            self._caps[w3c_name] = value
        return self 
```

So yes, I should be passing the value of `ADB_EXEC_TIMEOUT` below is the correct one.

```
class AdbExecTimeoutOption(SupportsCapabilities):
    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
    
    _transform_duration_get = DurationTransformer.transform_duration_get
    _transform_duration_set = DurationTransformer.transform_duration_set
    
    adb_exec_timeout = OptionsDescriptor(ADB_EXEC_TIMEOUT, _transform_duration_get, _transform_duration_set)
```

In the corrected version above , I have remove the quotes from `ADB_EXEC_TIMEOUT` while calling `OptionsDescriptor`. Now the actual value of `cls` variable `ADB_EXEC_TIMEOUT` will be passed, which is `adbExecTimeout`. I will change this in all other options.

mykola-mokhnach commented on appium/options/android/common/avd/avd_options.py:
```diff
@@ -0,0 +1,210 @@
+
+class AvdReadyTimeoutOption(SupportsCapabilities):
+    AVD_READY_TIMEOUT = 'avdReadyTimeout'
+    avd_ready_timeout = OptionsDescriptor('AVD_READY_TIMEOUT')
```
Transformers must be applied to all options where they are needed

sandeepsuryaprasad replied:
> Transformers must be applied to all options where they are needed

Yes. I have missed it. Thanks for letting me know. I will make all the changes, I will review it once again with the current release version and re-commit the code.

mykola-mokhnach commented on appium/options/base_options.py:
```diff
@@ -0,0 +1,34 @@
+from typing import Any, TypeVar, Generic, Type, Callable, Optional
```
not sure about the name of this module. it contains the descriptor class, but is called base_options

sandeepsuryaprasad replied:
Changed the name of the module to `base_options_descriptor`

mykola-mokhnach commented on appium/options/base_options.py:
```diff
@@ -0,0 +1,34 @@
+        value = obj.get_capability(self.name)
+        if self.tget:
+            # transform the value and then return
+            return self.tget(obj, value)
```
the signature of this function does not match to its type definition. Please run mypy as mentioned above

sandeepsuryaprasad commented:
I have incorporated review comments. Below are the changes incorporated.
1. Changed the name of `base_options.py` to `base_options_descriptor.py`
2. Removed `self` made the transformation functions `@staticmethod`
3. While creating instance of descriptor, now I am passing the actual value of `cls` variable which is in CamelCase
4. Instead of reassigning duration transformation functions that are imported from `transformation.py` to one more variable in `cls`, now I am directly passing transformation functions to descriptor.
5. Fixed type hints of transformation functions.

mykola-mokhnach commented on appium/options/transformers.py:
```diff
@@ -0,0 +1,13 @@
+from typing import Optional, Union
+
+
+class DurationTransformer:
```
I don't see a need to have these methods under a class. They could be simple functions

sandeepsuryaprasad replied:
Fixed. Moved functions out of the class.

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,367 @@
+
+class AdbExecTimeoutOption(SupportsCapabilities):
+    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
+    adb_exec_timeout = OptionsDescriptor(
```
generic type definitions are still missing for all OptionsDescriptor definitions. Did you run mypy on the source?

sandeepsuryaprasad replied:
Fixed. Added type hints to all OptionsDescriptor. attaching screen shot of mypy results.

 
<img width="802" alt="mypy" src="https://github.com/appium/python-client/assets/26169602/5099abbf-cb46-406d-867b-24fc922433ff">

mykola-mokhnach commented on appium/options/base_options_descriptor.py:
```diff
@@ -0,0 +1,34 @@
+
+    def __get__(self, obj: C, cls: Type[C]) -> R:
+        value = obj.get_capability(self.name)
+        if not self.tget is None:
```
prefer `is not None` syntax: https://stackoverflow.com/questions/2710940/python-if-x-is-not-none-or-if-not-x-is-none

sandeepsuryaprasad replied:
Fixed. Changed to `is not None`

mykola-mokhnach commented on appium/options/ios/safari/safari_options.py:
```diff
@@ -0,0 +1,293 @@
+    AUTOMATIC_INSPECTION = 'safari:automaticInspection'
+    automatic_inspection = OptionsDescriptor(AUTOMATIC_INSPECTION)
+    """
+    This capability instructs Safari to preload the Web Inspector and JavaScript
```
Please make sure docstrings follow Google style: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

sandeepsuryaprasad replied:
I have followed numpy style documentation. This is also very nice.
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy

mykola-mokhnach replied:
It is good, although for this particular project we have already agreed to use the Google one. I don't think it would make sense to have multiple mixed styles there

sandeepsuryaprasad replied:
I will change the docstrings according to Google's standard.

mykola-mokhnach commented:
Please also apply automated sorting and formatting (`make black`, `make isort`). 

And rename the PR itself to comply with https://www.conventionalcommits.org/en/v1.0.0/ spec

sandeepsuryaprasad commented:
applied isort and black formatting

<img width="535" alt="isort" src="https://github.com/appium/python-client/assets/26169602/44f559de-9f83-4cac-af67-fc4dcad135d1">

sandeepsuryaprasad commented:
> 

Fixed.

mykola-mokhnach commented on appium/options/android/common/adb/adb_options.py:
```diff
@@ -0,0 +1,367 @@
+class AdbExecTimeoutOption(SupportsCapabilities):
+    ADB_EXEC_TIMEOUT = "adbExecTimeout"
+    adb_exec_timeout = OptionsDescriptor[Optional[timedelta], Union[timedelta, int]]
+    (ADB_EXEC_TIMEOUT, transform_duration_get, transform_duration_set)
```
I assume this definition is not parsed correctly being split into two lines. Write it in a single line and then `make black` to wrap it automatically. Also applicable to other classes below

mykola-mokhnach commented on appium/options/android/common/avd/avd_options.py:
```diff
@@ -0,0 +1,217 @@
+
+class AvdEnvOption(SupportsCapabilities):
+    AVD_ENV = "avdEnv"
+    avd_env = OptionsDescriptor[Optional[Dict[str, str]], str](AVD_ENV)
```
the incoming value must also be dict

mykola-mokhnach commented on appium/options/android/common/avd/avd_options.py:
```diff
@@ -0,0 +1,217 @@
+
+
+class GpsEnabledOption(SupportsCapabilities):
+    GPS_ENABLED = "gpsEnabled"
```
double quotes here instead of single

mykola-mokhnach replied:
same below

sandeepsuryaprasad replied:
> double quotes here instead of single

I had used single quotes. but after formatting it has changed to double quotes..

mykola-mokhnach replied:
hm, this is some strange black config

@KazuCocoa does it do same changes for you by replacing single quotes with double ones?

sandeepsuryaprasad replied:
@mykola-mokhnach I saw the source code after formatting and I was surprised to see all single quotes converted to double quotes.. I thought it is taking as per config.. so I did not change it in source code..

KazuCocoa replied:
Hm, I thought we used single quote. Actually selenium Python binging required double quote though.

mykola-mokhnach commented on appium/options/android/common/locking/locking_options.py:
```diff
@@ -0,0 +1,165 @@
+
+class UnlockKeyOption(SupportsCapabilities):
+    UNLOCK_KEY = "unlockKey"
+    unlock_key = OptionsDescriptor[Optional[str], bool](UNLOCK_KEY)
```
is the incoming value of type boolean for this cap?

sandeepsuryaprasad replied:
My mistake.. I have fixed this. It should be a `str`

mykola-mokhnach commented on appium/options/android/common/mjpeg/mjpeg_screenshot_url_option.py:
```diff
@@ -19,7 +19,7 @@
 
 from appium.options.common.supports_capabilities import SupportsCapabilities
```
should we also update this option and below ones?

sandeepsuryaprasad replied:
Yes I have changed it to use descriptor.

mykola-mokhnach commented on appium/options/android/espresso/activity_options_option.py:
```diff
@@ -19,7 +19,7 @@
 
 from appium.options.common.supports_capabilities import SupportsCapabilities
```
should below option also be refactored?

sandeepsuryaprasad replied:
I have refactored 
`ActivityOptionsOption`
`AppLocaleOption`
`IntentOptionsOption`
`ShowGradleLogOption`

mykola-mokhnach replied:
I assume it should also be moved to a common module like other options which were previously in separate modules

mykola-mokhnach commented on appium/options/common/base.py:
```diff
@@ -20,16 +20,18 @@
-APPIUM_PREFIX = 'appium:'
-T = TypeVar('T', bound='AppiumOptions')
-PLATFORM_NAME = 'platformName'
+APPIUM_PREFIX = "appium:"
```
single quotes

mykola-mokhnach commented on appium/options/ios/xcuitest/app/app_options.py:
```diff
@@ -0,0 +1,115 @@
+
+class LocalizableStringsDirOption(SupportsCapabilities):
+    LOCALIZABLE_STRINGS_DIR = "localizableStringsDir"
+    localizable_strings_dir = OptionsDescriptor[Optional[str], str]
```
incorrect wrapping here

mykola-mokhnach commented on appium/options/ios/xcuitest/other/other_options.py:
```diff
@@ -0,0 +1,166 @@
+        Optional[Union[Dict[str, timedelta], timedelta]],
+        Union[Dict[str, timedelta], timedelta, int],
+    ]
+    (COMMAND_TIMEOUTS, transform_timeout_get, transform_timeout_set)
```
same as above

mykola-mokhnach commented on appium/options/mac/mac2/mac_options.py:
```diff
@@ -0,0 +1,217 @@
+class ServerStartupTimeoutOption(SupportsCapabilities):
+    SERVER_STARTUP_TIMEOUT = "serverStartupTimeout"
+    server_startup_timeout = OptionsDescriptor[Optional[timedelta], Union[int, timedelta]]
+    (SERVER_STARTUP_TIMEOUT, transform_duration_get, transform_duration_set)
```
same as above

mykola-mokhnach commented on appium/options/windows/windows/windows_options.py:
```diff
@@ -0,0 +1,195 @@
+class WaitForAppLaunchOption(SupportsCapabilities):
+    WAIT_FOR_APP_LAUNCH = "ms:waitForAppLaunch"
+    wait_for_app_launch = OptionsDescriptor[Optional[timedelta], Union[timedelta, int]](
+        WAIT_FOR_APP_LAUNCH, tset=transform_duration_set
```
why no getter is defined?

sandeepsuryaprasad replied:
fixed. It was over-looked.

mykola-mokhnach commented on test/unit/junit.xml:
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite name="pytest" errors="0" failures="0" skipped="0" tests="122" time="2.858" timestamp="2023-07-03T08:07:30.345923" hostname="Sandeeps-MacBook-Pro.local"><testcase classname="unit.webdriver.app_test.TestWebDriverApp" name="test_reset" time="0.017" /><testcase classname="unit.webdriver.app_test.TestWebDriverApp" name="test_install_app" tim ...
```
should not be under source control

KazuCocoa commented:
Checked that our black-related commands in pre-commit-config.yaml and Makefile give `-S`, so basically it should not raise any errors for the quote. So basically autoformat also should not work by our pre-commit-config.yaml. CI was also ok.

>   -S, --skip-string-normalization
>                                  Don't normalize string quotes or prefixes.

sandeepsuryaprasad commented:
I have pushed the latest code by fixing all review comments. (docstrings in Google's style yet to be applied)

mykola-mokhnach commented on appium/options/android/common/other/disable_suppress_accessibility_service_option.py:
```diff
@@ -19,7 +19,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-DISABLE_SUPPRESS_ACCESSIBILITY_SERVICE = 'disableSuppressAccessibilityService'
+DISABLE_SUPPRESS_ACCESSIBILITY_SERVICE = "disableSuppressAccessibilityService"
```
this option is still not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/common/other/user_profile_option.py:
```diff
@@ -19,7 +19,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-USER_PROFILE = 'userProfile'
+USER_PROFILE = "userProfile"
```
this option is still not up to date

sandeepsuryaprasad replied:
Updated.

mykola-mokhnach commented on appium/options/android/espresso/espresso_build_config_option.py:
```diff
@@ -20,7 +20,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-ESPRESSO_BUILD_CONFIG = 'espressoBuildConfig'
+ESPRESSO_BUILD_CONFIG = "espressoBuildConfig"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/espresso/espresso_server_launch_timeout_option.py:
```diff
@@ -20,7 +20,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-ESPRESSO_SERVER_LAUNCH_TIMEOUT = 'espressoServerLaunchTimeout'
+ESPRESSO_SERVER_LAUNCH_TIMEOUT = "espressoServerLaunchTimeout"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/espresso/force_espresso_rebuild_option.py:
```diff
@@ -19,7 +19,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-FORCE_ESPRESSO_REBUILD = 'forceEspressoRebuild'
+FORCE_ESPRESSO_REBUILD = "forceEspressoRebuild"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/uiautomator2/disable_window_animation_option.py:
```diff
@@ -19,7 +19,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-DISABLE_WINDOWS_ANIMATION = 'disableWindowAnimation'
+DISABLE_WINDOWS_ANIMATION = "disableWindowAnimation"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/uiautomator2/mjpeg_server_port_option.py:
```diff
@@ -19,7 +19,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-MJPEG_SERVER_PORT = 'mjpegServerPort'
+MJPEG_SERVER_PORT = "mjpegServerPort"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/uiautomator2/skip_device_initialization_option.py:
```diff
@@ -19,7 +19,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-SKIP_DEVICE_INITIALIZATION = 'skipDeviceInitialization'
+SKIP_DEVICE_INITIALIZATION = "skipDeviceInitialization"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/uiautomator2/skip_server_installation_option.py:
```diff
@@ -19,7 +19,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-SKIP_SERVER_INSTALLATION = 'skipServerInstallation'
+SKIP_SERVER_INSTALLATION = "skipServerInstallation"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/uiautomator2/uiautomator2_server_install_timeout_option.py:
```diff
@@ -20,7 +20,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-UIAUTOMATOR2_SERVER_INSTALL_TIMEOUT = 'uiautomator2ServerInstallTimeout'
+UIAUTOMATOR2_SERVER_INSTALL_TIMEOUT = "uiautomator2ServerInstallTimeout"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/uiautomator2/uiautomator2_server_launch_timeout_option.py:
```diff
@@ -20,7 +20,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-UIAUTOMATOR2_SERVER_LAUNCH_TIMEOUT = 'uiautomator2ServerLaunchTimeout'
+UIAUTOMATOR2_SERVER_LAUNCH_TIMEOUT = "uiautomator2ServerLaunchTimeout"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/android/uiautomator2/uiautomator2_server_read_timeout_option.py:
```diff
@@ -20,7 +20,7 @@
 from appium.options.common.supports_capabilities import SupportsCapabilities
 
-UIAUTOMATOR2_SERVER_READ_TIMEOUT = 'uiautomator2ServerReadTimeout'
+UIAUTOMATOR2_SERVER_READ_TIMEOUT = "uiautomator2ServerReadTimeout"
```
the option is not up to date

sandeepsuryaprasad replied:
Updated

mykola-mokhnach commented on appium/options/common/common_options.py:
```diff
@@ -0,0 +1,644 @@
+    def transfrom_command_timeout_set(value: Union[timedelta, int]) -> Union[int, float]:
+        return value.total_seconds() if isinstance(value, timedelta) else value
+
+    new_command_timeout = OptionsDescriptor[Optional[timedelta], Union[timedelta, int]]
```
the type definition for transformers differers from the actual one

mykola-mokhnach commented on appium/options/common/common_options.py:
```diff
@@ -0,0 +1,644 @@
+
+class PrintPageSourceOnFindFailureOption(SupportsCapabilities):
+    PRINT_PAGE_SOURCE_ON_FIND_FAILURE = "printPageSourceOnFindFailure"
+    print_page_source_on_find_failure = OptionsDescriptor[Optional[bool], bool]
```
this definition must be properly formatted

sandeepsuryaprasad replied:
Fixed

mykola-mokhnach commented on appium/options/ios/xcuitest/simulator/simulator_options.py:
```diff
@@ -0,0 +1,579 @@
+class ForceSimulatorSoftwareKeyboardPresenceOption(SupportsCapabilities):
+    FORCE_SIMULATOR_SOFTWARE_KEYBOARD_PRESENCE = "forceSimulatorSoftwareKeyboardPresence"
+    force_simulator_software_keyboard_presence = OptionsDescriptor[Optional[bool], bool]
+    (FORCE_SIMULATOR_SOFTWARE_KEYBOARD_PRESENCE)
```
this line must be formatted properly

mykola-mokhnach commented on appium/options/ios/xcuitest/simulator/simulator_options.py:
```diff
@@ -0,0 +1,579 @@
+        return json.dumps(value, ensure_ascii=False)
+
+    permissions = OptionsDescriptor[Optional[Dict[str, Dict[str, str]]], Dict[str, Dict[str, str]]]
+    (PERMISSIONS, transform_get, transform_set)
```
this definition must be properly formatted

mykola-mokhnach commented on appium/options/ios/xcuitest/simulator/simulator_options.py:
```diff
@@ -0,0 +1,579 @@
+class SimulatorPasteboardAutomaticSyncOption(SupportsCapabilities):
+    SIMULATOR_PASTEBOARD_AUTOMATIC_SYNC = "simulatorPasteboardAutomaticSync"
+    simulator_pasteboard_automatic_sync = OptionsDescriptor[Optional[bool], bool]
+    (SIMULATOR_PASTEBOARD_AUTOMATIC_SYNC)
```
this definition must be properly formatted

mykola-mokhnach commented on appium/options/ios/xcuitest/simulator/simulator_options.py:
```diff
@@ -0,0 +1,579 @@
+class SimulatorStartupTimeoutOption(SupportsCapabilities):
+    SIMULATOR_STARTUP_TIMEOUT = "simulatorStartupTimeout"
+    simulator_startup_timeout = OptionsDescriptor[Optional[timedelta], Union[timedelta, int]]
+    (SIMULATOR_STARTUP_TIMEOUT, transform_duration_get, transform_duration_set)
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+from appium.options.common.supports_capabilities import SupportsCapabilities
+
+
+class WDATransformTimeout:
```
I don't see a need to put these transformers into a class

sandeepsuryaprasad replied:
moved transformer functions out of class.

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+class AllowProvisioningDeviceRegistrationOption(SupportsCapabilities):
+    ALLOW_PROVISIONING_DEVICE_REGISTRATION = "allowProvisioningDeviceRegistration"
+    allow_provisioning_device_registration = OptionsDescriptor[Optional[bool], bool]
+    (ALLOW_PROVISIONING_DEVICE_REGISTRATION)
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+        Optional[Dict[str, Union[List[str], Dict[str, str]]]],
+        Dict[str, Union[List[str], Dict[str, str]]],
+    ]
+    (PROCESS_ARGUMENTS)
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+class WaitForIdleTimeoutOption(SupportsCapabilities):
+    WAIT_FOR_IDLE_TIMEOUT = "waitForIdleTimeout"
+    wait_for_idle_timeout = OptionsDescriptor[Optional[timedelta], Union[timedelta, float]]
+    (
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+class WdaConnectionTimeoutOption(SupportsCapabilities):
+    WDA_CONNECTION_TIMEOUT = "wdaConnectionTimeout"
+    wda_connection_timeout = OptionsDescriptor[Optional[timedelta], Union[timedelta, int]]
+    (
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+class WdaEventloopIdleDelayOption(SupportsCapabilities):
+    WDA_EVENTLOOP_IDLE_DELAY = "wdaEventloopIdleDelay"
+    wda_eventloop_idle_delay = OptionsDescriptor[Optional[timedelta], Union[timedelta, float]]
+    (
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+class WdaLaunchTimeoutOption(SupportsCapabilities):
+    WDA_LAUNCH_TIMEOUT = "wdaLaunchTimeout"
+    wda_launch_timeout = OptionsDescriptor[Optional[timedelta], Union[timedelta, int]]
+    (
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+
+class WDATransformTimeout:
+    @staticmethod
+    def transform_wda_timeout_get(value: Optional[int]) -> Optional[timedelta]:
```
Should be the value of type float?

mykola-mokhnach commented on appium/options/ios/xcuitest/wda/wda_options.py:
```diff
@@ -0,0 +1,1071 @@
+        return None if value is None else timedelta(seconds=value)
+
+    @staticmethod
+    def transform_wda_timeout_set(value: Union[timedelta, int]) -> Union[int, float]:
```
should be the value of type float as well?

Some WDA caps are different because they accept float seconds instead of milliseconds and there should be a way to set float values there and get timedeltas in response

mykola-mokhnach replied:
also any int number is a valid float  value

mykola-mokhnach commented on appium/options/ios/xcuitest/webview/webview_options.py:
```diff
@@ -0,0 +1,533 @@
+class SafariLogAllCommunicationHexDumpOption(SupportsCapabilities):
+    SAFARI_LOG_ALL_COMMUNICATION_HEX_DUMP = "safariLogAllCommunicationHexDump"
+    safari_log_all_communication_hex_dump = OptionsDescriptor[Optional[bool], bool]
+    (SAFARI_LOG_ALL_COMMUNICATION_HEX_DUMP)
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/webview/webview_options.py:
```diff
@@ -0,0 +1,533 @@
+class SafariLogAllCommunicationOption(SupportsCapabilities):
+    SAFARI_LOG_ALL_COMMUNICATION = "safariLogAllCommunication"
+    safari_log_all_communication = OptionsDescriptor[Optional[bool], bool]
+    (SAFARI_LOG_ALL_COMMUNICATION)
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/webview/webview_options.py:
```diff
@@ -0,0 +1,533 @@
+class SafariOpenLinksInBackgroundOption(SupportsCapabilities):
+    SAFARI_OPEN_LINKS_IN_BACKGROUND = "safariOpenLinksInBackground"
+    safari_open_links_in_background = OptionsDescriptor[Optional[bool], bool]
+    (SAFARI_OPEN_LINKS_IN_BACKGROUND)
```
same as above

mykola-mokhnach commented on appium/options/ios/xcuitest/webview/webview_options.py:
```diff
@@ -0,0 +1,533 @@
+class SafariSocketChunkSizeOption(SupportsCapabilities):
+    SAFARI_SOCKET_CHUNK_SIZE = "safariSocketChunkSize"
+    safari_socket_chunk_size = OptionsDescriptor[Optional[int], int]
+    (SAFARI_SOCKET_CHUNK_SIZE)
```
same

mykola-mokhnach commented on appium/options/ios/xcuitest/webview/webview_options.py:
```diff
@@ -0,0 +1,533 @@
+class SafariWebInspectorMaxFrameLengthOption(SupportsCapabilities):
+    SAFARI_WEB_INSPECTOR_MAX_FRAME_LENGTH = "safariWebInspectorMaxFrameLength"
+    safari_web_inspector_max_frame_length = OptionsDescriptor[Optional[int], int]
+    (SAFARI_WEB_INSPECTOR_MAX_FRAME_LENGTH)
```
same

mykola-mokhnach commented on appium/options/windows/windows/windows_options.py:
```diff
@@ -0,0 +1,204 @@
+class WaitForAppLaunchOption(SupportsCapabilities):
+    WAIT_FOR_APP_LAUNCH = "ms:waitForAppLaunch"
+
+    @staticmethod
```
I have already seen a duplicated transformer definition in WDA caps (e.g. allows to provide float seconds)

sandeepsuryaprasad replied:
since it is used in more than one capability moved the function to `transformers.py` 
`transform_duration_in_seconds_get` and `transform_duration_in_seconds_set`

mykola-mokhnach commented on appium/options/ios/xcuitest/webview/webview_options.py:
here as well

mykola-mokhnach commented on appium/options/ios/xcuitest/webview/webview_options.py:
and here

mykola-mokhnach commented on appium/options/ios/xcuitest/webview/webview_options.py:
the docstring seems to be cut off. What is the measurement unit and the default value?

sandeepsuryaprasad commented:
@mykola-mokhnach I hope I have fixed all most all the review comments, please feel free to let me know in case anything needs to be modified.

KazuCocoa commented on test/unit/junit.xml:
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite name="pytest" errors="0" failures="0" skipped="0" tests="122" time="5.144" timestamp="2023-07-03T22:03:51.611198" hostname="Sandeeps-MacBook-Pro.local"><testcase classname="unit.webdriver.app_test.TestWebDriverApp" name="test_reset" time="0.020" /><testcase classname="unit.webdriver.app_test.TestWebDriverApp" name="test_install_app" tim ...
```
should be excluded?
